include/swift/AST/Decl.h

```diff
@@ -666,6 +666,13 @@ class alignas(1 << DeclAlignInBits) Decl {
   void operator=(const Decl&) = delete;
   SourceLoc getLocFromSource() const;
 
+  struct CachedExternalSourceLocs {
+    SourceLoc Loc;
+    SourceLoc StartLoc;
+    SourceLoc EndLoc;
+  };
+  mutable Optional<CachedExternalSourceLocs> SerializedLocsCache;
```

Comment:
:-( It's unfortunate that this bloats the size of every Decl. Maybe we should just always have these three source locations, and bits to say whether they're computed, and then we can cache the normal lookups too? Or have them behind an indirection allocated on the ASTContext, so that it's only one word instead of three, and only if there are actual valid locations.